File: src/main/java/com/xu/music/player/wrapper/BasicWrapper.java
package com.xu.music.player.wrapper;

import cn.hutool.core.util.StrUtil;
import com.xu.music.player.hander.DataBaseError;
import com.xu.music.player.wrapper.sql.SqlCommand;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;

/**
 * SQL Wrapper 的公共状态与参数收集逻辑。
 */
public class BasicWrapper<T> {

    private static final Pattern UPPER_CASE = Pattern.compile("[A-Z]");
    private static final Pattern IDENTIFIER = Pattern.compile("[A-Za-z_][A-Za-z0-9_]*");

    protected Class<T> bean;
    protected String table;
    protected String[] field;
    protected final List<String> conditions = new ArrayList<>();
    protected final List<Object> parameters = new ArrayList<>();

    protected void addCondition(String sql, Object... values) {
        conditions.add(" and " + sql);
        Collections.addAll(parameters, values);
    }

    protected SqlCommand command(String baseSql, List<Object> leadingParameters) {
        var sql = new StringBuilder(baseSql);
        conditions.forEach(sql::append);

        var allParameters = new ArrayList<>(leadingParameters);
        allParameters.addAll(parameters);
        return new SqlCommand(sql.toString(), allParameters);
    }

    protected static String requireIdentifier(String name) {
        if (name == null || !IDENTIFIER.matcher(name).matches()) {
            throw new DataBaseError("非法 SQL 标识符: " + name);
        }
        return name;
    }

    protected String dealField(String name) {
        requireIdentifier(name);
        var matcher = UPPER_CASE.matcher(name);
        var result = new StringBuilder();
        while (matcher.find()) {
            matcher.appendReplacement(result, "_" + matcher.group().toLowerCase(Locale.ROOT));
        }
        matcher.appendTail(result);
        var fieldName = result.toString();
        return StrUtil.equalsAnyIgnoreCase(fieldName, "index") ? "`" + fieldName + "`" : fieldName;
    }
}

File: src/main/kotlin/com/xu/music/player/wrapper/BasicWrapper.kt
package com.xu.music.player.wrapper

import java.lang.reflect.Modifier
import java.text.SimpleDateFormat
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Date
import java.util.Locale

/** 数据库操作的字段与值转换。 */
open class BasicWrapper<T> {
    protected var last: String = ""
    protected lateinit var bean: Class<T>
    protected lateinit var table: String
    protected var fields: List<String> = listOf("*")
    protected val condition = mutableListOf<String>()

    protected fun dealValue(values: List<Any?>): String = values.joinToString(",") { dealValue(it) }

    protected fun dealValue(value: Any?): String = when (value) {
        null -> "null"
        is Date -> "'${SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(value)}'"
        is LocalDateTime -> "'${value.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))}'"
        is String -> "'${value.replace("'", "''")}'"
        else -> value.toString()
    }

    protected fun dealField(name: String): String {
        val column = Regex("[A-Z]").replace(name) { "_" + it.value.lowercase(Locale.ROOT) }
        return if (column.equals("index", ignoreCase = true)) "`$column`" else column
    }

    /** 忽略静态字段和空值，保留原有仅写入非空属性的约定。 */
    protected fun values(data: Any): List<Pair<String, Any>> = data.javaClass.declaredFields
        .filterNot { Modifier.isStatic(it.modifiers) || it.isSynthetic }
        .mapNotNull { field ->
            field.isAccessible = true
            field.get(data)?.let { dealField(field.name) to it }
        }

    protected fun insertSql(data: Any): String {
        val values = values(data)
        return "insert into $table(${values.joinToString(", ") { it.first }}) " +
            "values(${dealValue(values.map { it.second })})"
    }

    protected fun whereSql(): String = " where 1 = 1 " + condition.joinToString(" ") + last
}
